### PR TITLE
Add default cache headers

### PR DESCRIPTION
**Pull request by jpmartinspt**
# Done
Add a default strategy to add cache headers to all flask-base responses.

# QA
- Verify the tests I wrote make sense
- Make sure the tests pass

# Issues
https://github.com/canonical-web-and-design/web-squad/issues/2312

### CODE REVIEW

File: canonicalwebteam/flask_base/request.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from flask import request, session
+
+
+def set_cache_control_headers(response):
+    print(request.path)
```
Comment (tbille) [inline canonicalwebteam/flask_base/request.py]: oups print

File: .github/workflows/flask-base.yml [hunk 2]
```diff
@@ -33,9 +33,7 @@ jobs:
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
```
Comment (nottrobin) [inline .github/workflows/flask-base.yml]: Won't this run with python2?

File: canonicalwebteam/flask_base/request.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+from flask import request, session
+
+
+def set_cache_control_headers(response):
```
Comment (nottrobin) [inline canonicalwebteam/flask_base/request.py]: Couldn't we just do this in app.py? I'm not a huge fan of creating a new file per function.

File: tests/test_app/webapp/app.py [hunk 4]
```diff
@@ -19,6 +19,24 @@ def create_test_app():
     def page():
         return "page"
 
+    @app.route("/auth")
```
Comment (nottrobin) [inline tests/test_app/webapp/app.py]: :heart:

File: .github/workflows/flask-base.yml (hunk 2 above)
Comment (jpmartinspt) [reply .github/workflows/flask-base.yml]: AFAIK the setup-python action above, makes sure I have the correct python version.

**Comment by nottrobin** [comment]
@jpmartinspt I've been schooled on Flask's session defaults - this might have some implications for this PR: https://github.com/pallets/flask/issues/3423.

Perhaps we should be using `session.accessed`? And also perhaps the `Vary: Cookie` default is actually sufficient and so we should leave any extra headers to the sites to add themselves (e.g. ubuntu.com/advantage still needs `private`, presumably).

What examples did we see where caching has bitten us? Do we actually have a reason to believe that the `Vary: Cookie` solution is unsafe? @tbille do you remember that issue we had with snapcraft a long time back - how did we fix that?

**Comment by jpmartinspt** [comment]
@nottrobin @tbille  what about starting like this and slowly customizing things across the sites to see how things behave and when we are happy we can come back to remove stuff from flask-base?

**Comment by nottrobin** [comment]
What would be the trigger to make us come back and revisit? I suspect we might never do that because it worked, but be hurting user performance by cache busting much more than we need to across all our sites.

It seems to me like we should err on the other side - since flask-base has wide ranging impact we should be cautious with it. Do the minimum that we know is required and revisit if it turn out not to be sufficient.

So how about we don't do the `private` stuff for now, but test that `Vary: cookie` is indeed set by flask and that that genuinely does bust the cache for users? Which would mean this pr becomes just about default caching headers & stale while revalidate.

**Comment by jpmartinspt** [comment]
Makes sense, updated the code and the test.

**Pull request merged**